File: SWEA/D2/1954. 달팽이 숫자/달팽이 숫자.py
# 델타 탐색을 위한 방향 인덱스
dx = [0, 1, 0, -1] # 동 남 서 북
dy = [1, 0, -1, 0]

t = int(input())
for tc in range(1, t + 1):
    n = int(input())
    snail = [[0] * n for _ in range(n)] # 달팽이 2차원 리스트

    x, y, idx = 0, 0, 0 # 초기값 세팅
    for i in range(1, n ** 2 + 1):
        snail[x][y] = i
        
        # 달팽이가 이동합니다
        nx = x + dx[idx]
        ny = y + dy[idx]
        
        # 다음 인덱스가 범위 내이고, 아직 지나가지 않은 칸이라면 이동
        if 0 <= nx < n and 0 <= ny < n and not snail[nx][ny]:
            x, y = nx, ny
        else:
            # 방향 인덱스는 3을 넘지 않도록 4로 나눈 나머지로 업뎃!
            idx = (idx + 1) % 4
            x += dx[idx]
            y += dy[idx]

    print(f'#{tc}')
    for i in range(n):
        print(*snail[i])

File: SWEA/D2/1954. 달팽이 숫자/달팽이 숫자.java
//import java.io.FileInputStream;
import java.util.Scanner;

public class Solution {

	public static void main(String[] args) throws Exception {
//		System.setIn(new FileInputStream("src/input.txt"));
		Scanner sc = new Scanner(System.in);
		
		int[] dx = {0, 1, 0, -1};
		int[] dy = {1, 0, -1, 0};
		
		int t = sc.nextInt();
		for (int tc = 1; tc <= t; tc++) {
			int n = sc.nextInt();
			int[][] snail = new int[n][n];
			
			int x = 0;
			int y = 0;
			int idx = 0;
			for (int i = 1; i <= n * n; i++) {
				snail[x][y] = i;
				
				int nx = x + dx[idx];
				int ny = y + dy[idx];
				
				if ((0 <= nx && nx < n) && (0 <= ny && ny < n) && snail[nx][ny] == 0) {
					x = nx;
					y = ny;
				} else {
					idx = (idx + 1) % 4;
					x += dx[idx];
					y += dy[idx];
				}
			}
			
			System.out.printf("#%d%n", tc);
			for (int i = 0; i < n; i++) {
				for (int j = 0; j < n; j++) {
					System.out.print(snail[i][j] + " ");
				}
				System.out.println();
			}
		}
	}

}